wsmaster/codenvy-hosted-api-workspace/src/main/java/com/codenvy/api/workspace/server/jpa/RemovePermissionsOnLastUserRemovedEventSubscriber.java

```diff
@@ -0,0 +1,105 @@
+
+import com.codenvy.api.permission.server.model.impl.AbstractPermissions;
+import com.codenvy.api.permission.server.spi.PermissionsDao;
+import com.codenvy.api.workspace.server.recipe.RecipeDomain;
+import com.codenvy.api.workspace.server.stack.StackDomain;
+
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.core.notification.EventSubscriber;
+import org.eclipse.che.api.machine.server.jpa.JpaRecipeDao;
+import org.eclipse.che.api.user.server.event.BeforeUserRemovedEvent;
+import org.eclipse.che.api.user.server.model.impl.UserImpl;
+import org.eclipse.che.api.workspace.server.jpa.JpaStackDao;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.util.Set;
+
+import static com.codenvy.api.permission.server.AbstractPermissionsDomain.SET_PERMISSIONS;
+import static java.lang.String.format;
+
+/**
+ * Listens for {@link UserImpl} removal events, and checks if the removing user is the last who have "setPermissions"
+ * role to any of the permission domain, and if it is, then removes domain entity itself.
+ *
+ * @author Max Shaposhnik
+ */
+public class RemovePermissionsOnLastUserRemovedEventSubscriber implements EventSubscriber<BeforeUserRemovedEvent> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RemovePermissionsOnLastUserRemovedEventSubscriber.class);
+
+    @Inject
+    Set<PermissionsDao<? extends AbstractPermissions>> storages;
+
+    @Inject
+    JpaStackDao stackDao;
+
+    @Inject
+    JpaRecipeDao recipeDao;
+
+    @Inject
+    private EventService eventService;
+
+    @PostConstruct
+    public void subscribe() {
+        eventService.subscribe(this);
+    }
+
+    @PreDestroy
+    public void unsubscribe() {
+        eventService.unsubscribe(this);
+    }
+
+    @Override
+    public void onEvent(BeforeUserRemovedEvent event) {
+        try {
+            for (PermissionsDao<? extends AbstractPermissions> storage : storages) {
+                for (AbstractPermissions permissions : storage.getByUser(event.getUser().getId())) {
+                    // This method can  potentially be source of race conditions,
+                    // e.g. when performing search by permissions, another thread can add/or remove another setPermission,
+                    // so appropriate domain object (stack or recipe) will not be deleted, or vice versa,
+                    // deleted when it's not required anymore.
+                    // As a result, a solitary objects may be present in the DB.
+                    if (storage.getByInstance(permissions.getInstanceId())
+                               .stream()
+                               .noneMatch(permissions1 -> permissions1.getActions().contains(SET_PERMISSIONS) &&
+                                                          !permissions1.getUserId().equals(event.getUser().getId()))) {
```

Comment:
I don't get why do we have to check whether user is the same when we get permissions by user `storage.getByUser(event.getUser().getId())`, do i miss something ?